Multitouch Support/Native/VoodooI2CNativeEngine.cpp

```diff
@@ -137,14 +139,32 @@ bool VoodooI2CNativeEngine::isForceClickEnabled() {
         lastForceClickPropertyUpdateTime = now_abs;
         return lastIsForceClickEnabled;
     }
-
+    
     // Blocking reading here takes about 10 microseconds
     OSDictionary* dict = OSDynamicCast(OSDictionary, this->getProperty("MultitouchPreferences", gIOServicePlane, kIORegistryIterateRecursively));
     if (!dict) {
         return lastIsForceClickEnabled;
     }
     if (OSCollectionIterator* i = OSCollectionIterator::withCollection(dict)) {
         while (OSSymbol* key = OSDynamicCast(OSSymbol, i->getNextObject())) {
+            // If tap to click (property "Clicking") is disabled,
+            // then we should forbid force clicking,
+            // because it makes impossible to perform a non-force click
+            if (key->isEqualTo("Clicking")) {
+                OSBoolean *boolValue;
+                OSNumber *numValue;
+                if ((boolValue = OSDynamicCast(OSBoolean, dict->getObject(key))) != NULL) {
+                    if (!boolValue->getValue()) {
+                        lastIsForceClickEnabled = FALSE;
+                        break;
+                    }
+                }
+                // is it needed? in default properties it's number, not boolean
+                else if ((numValue = OSDynamicCast(OSNumber, dict->getObject(key))) != NULL && !numValue->unsigned64BitValue()) {
```

Comment:
Stylistic: This is also rather hard to read, can we please pull the assignment out of the if condition?

Reply:
I can pull it out of condition, but it will increase the nesting level, because the cast should be executed only if the previous cast to boolean did not succeed. (In fact, I have originally written it that way).